e2e-tests/tests/util.ts

```diff
@@ -186,3 +199,149 @@ export async function extractRevertReason(context: { web3: Web3 }, transactionHa
 		return context.web3.utils.hexToUtf8("0x" + reasonHex.slice(64)).trim(); // skip the padding and remove return carriage
 	}
 }
+
+// Generic function to send a transaction and wait for finalization
+export async function sendTxAndWaitForFinalization(
+	api: ApiPromise,
+	tx: SubmittableExtrinsic<ApiTypes>,
+	signer: KeyringPair,
+	waitNBlocks = 10
+) {
+	return new Promise((resolve, reject) => {
+		try {
+			let blockCount = 0;
+
+			const onStatusChange = async (result: SubmittableResult) => {
+				const { status, events, dispatchError } = result;
+				console.log("Transaction status:", status.type);
+
+				// Additional event info
+				if (events && events.length > 0) {
+					events.forEach(({ event: { method, section, data } }) => {
+						console.log(`Event: ${section}.${method} - Data: ${data.toString()}`);
+					});
+				}
+
+				if (dispatchError) {
+					console.error("Raw dispatch error:", dispatchError.toString());
+
+					if (dispatchError.isModule) {
+						const decoded = api.registry.findMetaError(dispatchError.asModule);
+						const { section, name, docs } = decoded;
+						console.error(`Transaction failed with error: ${section}.${name}`);
+						console.error(`Error documentation: ${docs.join(" ")}`);
+						reject(new Error(`${section}.${name}: ${docs.join(" ")}`));
+					} else {
+						console.error(`Transaction failed with error: ${dispatchError.toString()}`);
+						reject(new Error(dispatchError.toString()));
+					}
+					return;
+				}
+
+				if (status.isInBlock) {
+					console.log("Included at block hash", status.asInBlock.toHex());
+				} else if (status.isFinalized) {
+					console.log("Finalized block hash", status.asFinalized.toHex());
+					resolve(status.asFinalized.toHex());
+				} else if (status.isDropped || status.isInvalid || status.isUsurped) {
+					console.error("Transaction failed with status:", status.type);
+					// Start waiting for N blocks and re-check transaction status
+					const extrinsicHash = result.txHash.toHuman();
+					console.log(`Transaction is invalid. Waiting for ${waitNBlocks} blocks while rechecking status...`);
+
+					// Subscribe to new finalized blocks to re-check transaction status
+					const unsubscribeAll = await api.rpc.chain.subscribeFinalizedHeads(async (lastHeader) => {
+						blockCount++;
+						console.log(`Finalized Block #${lastHeader.number} received (${blockCount}/${waitNBlocks})`);
+
+						// Check if the transaction has been included in the block
+						const blockHash = lastHeader.hash;
+						const block = await api.rpc.chain.getBlock(blockHash);
+
+						let txIncluded = false;
+
+						for (const extrinsic of block.block.extrinsics) {
+							if (extrinsic.hash.toHex() === extrinsicHash) {
+								console.log(`Transaction included in block ${lastHeader.number}`);
+								txIncluded = true;
+								unsubscribeAll && unsubscribeAll();
```

Comment:
`unsubscribeAll` won't be undefined as this function is the callback passed to it, so it exists. Also call `unsubTx`

```suggestion
								unsubscribeAll();
								unsubTx();
```